Implement StreamingTimeArray and ReadOnlyArray

timholy opened the pull request:
Note that this PR is not complete because it's lacking tests; I'm interested in feedback about whether this is useful.

This adds a new ~~AbstractArray~~ type,  `StreamingTimeArray`, to represent array types possessing a time axis but for which switching time "slices" may have restrictions or may not be cheap.

At the moment I envision two uses for this type:
- [VideoIO](https://github.com/kmsquire/VideoIO.jl) (ping @kmsquire), where decompressing an AVI is done on a frame-by-frame basis; at the moment we don't have full "random access" to the stream, but even if we do some day, switching time slices is going to be expensive.
- A "demo" package [SpatiallyFilteredViews](https://github.com/timholy/SpatiallyFilteredViews.jl), which allows one to apply spatial filtering "on the fly" to a large movie. This package is incomplete, but it seems useful enough in my own work that I suspect I will finish it. If it's of general use I could move it to JuliaImages.

One specific thing I'd like feedback on is the degree to which these (and especially the latter) are an abuse of the "view" concept. It is a view in the sense that it's lazy and doesn't allocate a tremendous amount of memory. However, it doesn't follow the "changes in the view also generate changes in the parent." For that reason I also added a `ReadOnlyArray` wrapper, so that any attempt to set pixel values in the view will result in an error (of course users can still copy the data to any `Array` and then tweak to their heart's content).

timholy committed:
Implement StreamingTimeArray and ReadOnlyArray

timholy commented:
I've generalized this to support arbitrary streamed axes, and also to make it a concrete type rather than an abstract type. Now it's essentially a `ReadOnlyMappedArray` without the requirement that the mapping function be element-by-element.

With this, I don't think a separate SpatiallyFilteredViews package is necessary, since it can be implemented simply as

```julia
julia> f!(dest, A) = imfilter!(dest, A, KernelFactors.gaussian((1.5,1.5)), NA())
f! (generic function with 1 method)

julia> imgs = StreamingArray{RGB{Float64}}(f!, img, Axis{:time}())
```

To handle my concerns about abuse of `view` above, I've reworked the API around a `getindex!` method. If one wants to use this directly, the caller is responsible for managing the buffer. This should also eliminate potential worries about aliasing (e.g., `v1, v2 = view(S, Axis{:time}(1)), view(S, Axis{:time}(2))` would be a bug with the original API).

Evizero commented on src/ImageAxes.jl:
```diff
@@ -167,6 +178,159 @@ function Base.convert{Cdest<:Colorant,n,Csrc<:Colorant}(::Type{Array{Cdest,n}},
+    inds = findin(axisnames(axs.parameters...), streamingaxisnames(S))
+    Expr(:tuple, Expr[:(a[$i]) for i in inds]...)
+end
+@generated function _s!filter{N}(a, axs::NTuple{N,Axis}, S::StreamingArray)
```
probably not the kind of feedback you are looking for, but these `!` placements look very odd to me. It also seems that github has trouble with syntax highlighting them. Is this an emerging trend? How is the placement decided

timholy replied:
I agree, they're weird. These are the opposite axes of the ones that `sfilter` extracts, so the `!` is meant as "not." The function name can't start with `!`, otherwise it would be first.

Happy to hear better suggestions. Since it was internal I didn't sweat it too much but agree it's not pretty now.

Evizero replied:
Just to be clear, I am not suggesting that I think it has to be changed. I just found it surprising enough to point it out. 

How about `not_sfilter`. It's verbose but pretty clear. My initial thought was that the `!` was suggesting some mutation, but the placement was supposed to signal that something else than the first parameter is mutated.

timholy replied:
I hadn't thought about the mutation interpretation, definitely confusing. Using `not` seems reasonable to me.

timholy committed:
Make StreamingArray a concrete type

Emphasizes containment rather than subtyping

timholy committed:
StreamingArray->StreamingContainer

This allows us to avoid supporting `view`, for which an efficient implementation doesn't really exist without violating the contracts that `view` makes.

timholy committed:
Add 0.6 to the CI matrix

timholy commented:
I took a step even further, and decided to name this `StreamingContainer` and *not* make it an `AbstractArray` (but to give it much of the `AbstractArray` interface). I think this is more consistent with the spirit of the meaning of `AbstractArray`. This risks making it more difficult to use, but I think it's better to have to add methods than to have things misbehave due to inappropriate usages. In particular, `view`s of some streaming containers might behave really, really badly, and this prevents one from taking a view without writing a custom method for doing so.

I've taken off the RFC tag and,  barring objections, will merge fairly soon.

tkelman commented on src/ImageAxes.jl:
```diff
@@ -4,12 +4,23 @@ module ImageAxes
 
 using Base: @pure, tail
 using Reexport, Colors, SimpleTraits, MappedArrays
+using Compat
```
missing from REQUIRE (looks like you only need 0.17?)

tkelman commented on test/runtests.jl:
```diff
@@ -1,14 +1,25 @@
+end
+@test isempty(ambs)
 
 if VERSION < v"0.6.0-dev"
```
would it be better to be more conservative with this and set this to a known-working point, so it just doesn't test the inference on earlier versions as opposed to failing there?
